### ISSUE TITLE
Data corruption during a bitmap scan

### ISSUE THREAD

**Issue by braaatik** [greenplum-db/gpdb#11308]
### Greenplum version or build
PostgreSQL 9.4.24 (Greenplum Database 6.13.0 build dev) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39), 64-bit compiled
### OS version and uname -a
CentOS Linux 7 (Core)
Linux 3.10.0-1127.13.1.el7.x86_64
### Expected behavior
```
begin;
set local optimizer = off;

select distinct id
  from bug.simply_distributed_table
 where id = 146;
id
----------
146
```
### Actual behavior
Filter (id = 146) works randomly: well or partially
```
begin;
set local optimizer = off;

select distinct id
  from bug.simply_distributed_table
 where id = 146;
id
----------
112
125
146

select distinct id
  from bug.simply_distributed_table
 where id = 146;
id
----------
146
```
### Step to reproduce the behavior
It's a floating bug, but 3-5 attempts are enough for reproduction. All aspects are important: vacuum, complex distribution key in source table, data skew. Table with 20+ columns reproduced bug more often. Seems something wrong with VACUUM too, because the bug reproduced only when VACUUM works long (AO table segments recreation). But only every other time VACUUM works long.

DDL:
```
create schema bug;

create table bug.simply_distributed_table
(
  id integer
)
with (appendonly = true, orientation = column, compresstype = zstd, compresslevel = 9)
distributed by (id);

create index idx_simply_distributed_table__id on bug.simply_distributed_table using bitmap (id);


create table bug.complex_distributed_table
(
  id  integer,
  id2 integer
)
with (appendonly = true, orientation = column, compresstype = zstd, compresslevel = 9)
distributed by (id, id2);

insert into bug.complex_distributed_table(id, id2)
  with w(cnt) as (
    values (12),
           (14),
           (40),
           (45),
           (70),
           (76),
           (79),
           (94),
           (241),
           (298),
           (560),
           (575),
           (623),
           (701),
           (749),
           (780),
           (869),
           (900),
           (918),
           (1225),
           (1253),
           (1460),
           (1481),
           (1603),
           (2116),
           (2142),
           (2177),
           (2380),
           (2446),
           (2454),
           (2672),
           (2956),
           (3501),
           (3774),
           (4718),
           (6081),
           (6241),
           (6631),
           (7070),
           (7096),
           (7802),
           (8110),
           (8488),
           (8894),
           (9016),
           (9345),
           (9417),
           (9792),
           (10111),
           (10402),
           (10426),
           (10476),
           (10590),
           (11562),
           (12392),
           (13001),
           (13909),
           (14736),
           (14772),
           (15414),
           (16353),
           (17131),
           (18175),
           (18697),
           (20026),
           (20489),
           (20570),
           (21078),
           (21445),
           (22730),
           (23726),
           (24568),
           (24819),
           (26506),
           (27130),
           (29548),
           (29601),
           (29721),
           (30338),
           (30958),
           (31137),
           (31226),
           (32317),
           (33274),
           (33642),
           (35014),
           (35048),
           (35366),
           (35589),
           (35614),
           (38362),
           (39461),
           (40166),
           (40767),
           (41411),
           (46270),
           (53623),
           (60648),
           (63108),
           (63811),
           (63933),
           (64319),
           (66520),
           (66771),
           (71245),
           (72314),
           (78598),
           (84097),
           (86903),
           (93129),
           (106238),
           (116523),
           (116554),
           (129428),
           (131198),
           (149828),
           (153645),
           (163017),
           (164860),
           (175073),
           (177514),
           (187383),
           (210671),
           (214214),
           (244687),
           (249515),
           (266064),
           (283333),
           (285198),
           (315470),
           (341402),
           (344949),
           (377141),
           (407933),
           (519489),
           (520024),
           (598721),
           (635259),
           (689848),
           (827194),
           (882053),
           (953014),
           (960967),
           (1021011),
           (1341610),
           (1404254),
           (2134944),
           (2884251),
           (3027959),
           (6917419)
  )
  select wr.id, wr.id
    from (select row_number() over () as id,
                 w.cnt
            from w) wr
   cross join generate_series(1, wr.cnt) m;

analyze bug.complex_distributed_table;
```
Bug reproduction. Code from connection 2 must be executed only in parallel with `insert`:
```
--connection 1                                          --connection 2
vacuum full bug.simply_distributed_table;

begin;
  delete from bug.simply_distributed_table;

  insert into bug.simply_distributed_table(id)         do $$
    select c.id                                        declare
      from bug.complex_distributed_table c;              i int;
                                                         cnt bigint;
                                                       begin
                                                         set local optimizer = off;

                                                         for i in 1..50 loop
                                                           select count(distinct id)
                                                             into cnt
                                                             from bug.simply_distributed_table
                                                            where id = 146;

                                                           if cnt > 1 then
                                                             raise 'bug! (count: %)', cnt;
                                                           end if;
                                                         end loop;
                                                       end;
                                                       $$;
commit;
```

**Comment by JunfengYang** [comment]
When I trying to reproduce the issue, I encounter an assertion failure when executing the `select count(distinct id) from bug.simply_distributed_table where id = 146`.
```
#0  0x00007fec8f85403f in __GI___select (nfds=0, readfds=0x0, writefds=0x0, exceptfds=0x0, timeout=0x7fffa35564f0) at ../sysdeps/unix/sysv/linux/select.c:41
#1  0x0000557ecace6fef in pg_usleep (microsec=30000000) at pgsleep.c:56
#2  0x0000557ecabbed06 in elog_debug_linger (edata=0x557ecb2b3540 <errordata>) at elog.c:5261
#3  0x0000557ecabb47b0 in errfinish (dummy=0) at elog.c:670
#4  0x0000557ecabb2e33 in ExceptionalCondition (conditionName=0x557ecad4a593 "!(h->bytes_4_5 <= 0x8000)", errorType=0x557ecad4a56f "FailedAssertion", fileName=0x557ecad4a540 "../../../../src/include/access/appendonlytid.h",
    lineNumber=66) at assert.c:46
#5  0x0000557eca60fb83 in AOTupleIdGet_rowNum (h=0x7fffa355679a) at ../../../../src/include/access/appendonlytid.h:66
#6  0x0000557eca615687 in aocs_fetch (aocsFetchDesc=0x557ecb809d98, aoTupleId=0x7fffa355679a, slot=0x557ecb809c68) at aocsam.c:1322
#7  0x0000557eca7fcb67 in BitmapAppendOnlyNext (node=0x557ecb8b3588) at nodeBitmapHeapscan.c:797
#8  0x0000557eca7e7cec in ExecScanFetch (node=0x557ecb8b3588, accessMtd=0x557eca7fc5fb <BitmapAppendOnlyNext>, recheckMtd=0x557eca7fcc40 <BitmapHeapRecheck>) at execScan.c:84
#9  0x0000557eca7e7d57 in ExecScan (node=0x557ecb8b3588, accessMtd=0x557eca7fc5fb <BitmapAppendOnlyNext>, recheckMtd=0x557eca7fcc40 <BitmapHeapRecheck>) at execScan.c:132
#10 0x0000557eca7fccf2 in ExecBitmapHeapScan (node=0x557ecb8b3588) at nodeBitmapHeapscan.c:868
#11 0x0000557eca7d7e6b in ExecProcNode (node=0x557ecb8b3588) at execProcnode.c:1027
#12 0x0000557eca83d027 in agg_hash_initial_pass (aggstate=0x557ecb809120) at execHHashagg.c:1036
#13 0x0000557eca7f70e9 in ExecAgg (node=0x557ecb809120) at nodeAgg.c:1149
#14 0x0000557eca7d7fa6 in ExecProcNode (node=0x557ecb809120) at execProcnode.c:1093
#15 0x0000557eca82ecdb in execMotionSender (node=0x557ecb808aa0) at nodeMotion.c:295
#16 0x0000557eca82eb97 in ExecMotion (node=0x557ecb808aa0) at nodeMotion.c:252
#17 0x0000557eca7d8039 in ExecProcNode (node=0x557ecb808aa0) at execProcnode.c:1121
#18 0x0000557eca7d1d6b in ExecutePlan (estate=0x557ecb808758, planstate=0x557ecb808aa0, operation=CMD_SELECT, sendTuples=0 '\000', numberTuples=0, direction=ForwardScanDirection, dest=0x557ecb8c32e8) at execMain.c:2991
#19 0x0000557eca7ce105 in standard_ExecutorRun (queryDesc=0x557ecb87f9b8, direction=ForwardScanDirection, count=0) at execMain.c:935
#20 0x0000557eca7cdd24 in ExecutorRun (queryDesc=0x557ecb87f9b8, direction=ForwardScanDirection, count=0) at execMain.c:818
#21 0x0000557ecaa3135d in PortalRunSelect (portal=0x557ecb80a778, forward=1 '\001', count=0, dest=0x557ecb8c32e8) at pquery.c:1151
#22 0x0000557ecaa30f65 in PortalRun (portal=0x557ecb80a778, count=9223372036854775807, isTopLevel=1 '\001', dest=0x557ecb8c32e8, altdest=0x557ecb8c32e8,
    completionTag=0x7fffa3557000 <error: Cannot access memory at address 0x7fffa3557000>) at pquery.c:992
#23 0x0000557ecaa27ac5 in exec_mpp_query (query_string=0x557ecb80052e <error: Cannot access memory at address 0x557ecb80052e>, serializedQuerytree=0x0, serializedQuerytreelen=0,
    serializedPlantree=0x557ecb800593 <error: Cannot access memory at address 0x557ecb800593>, serializedPlantreelen=1687, serializedParams=0x0, serializedParamslen=0,
    serializedQueryDispatchDesc=0x557ecb800c2a <error: Cannot access memory at address 0x557ecb800c2a>, serializedQueryDispatchDesclen=184) at postgres.c:1367
#24 0x0000557ecaa2e593 in PostgresMain (argc=1, argv=0x557ecb813a90, dbname=0x557ecb7df5f8 <error: Cannot access memory at address 0x557ecb7df5f8>, username=0x557ecb7df5d8 <error: Cannot access memory at address 0x557ecb7df5d8>)
    at postgres.c:5388
#25 0x0000557eca98fba6 in BackendRun (port=0x557ecb80f4e0) at postmaster.c:4811
#26 0x0000557eca98f14f in BackendStartup (port=0x557ecb80f4e0) at postmaster.c:4468
#27 0x0000557eca98ac55 in ServerLoop () at postmaster.c:1948
#28 0x0000557eca98a01b in PostmasterMain (argc=5, argv=0x557ecb7dd490) at postmaster.c:1518
```

It breaks the assertion `Assert(h->bytes_4_5 <= 0x8000);` in function `AOTupleIdGet_rowNum`.

**Comment by braaatik** [comment]
We havent such problem. Seems like other bug. You can try to replace anonymous block by multiple manual code execution:
```
begin;
  set local optimizer = off;

  select distinct id
    from bug.simply_distributed_table
   where id = 146;
commit;
```

**Comment by JunfengYang** [comment]
@braaatik Is your build enable assert?
I assume this is the same issue since `AOTupleIdGet_rowNum` will return a wrong row number without the assert which may point to a visible ao tuple.

**Comment by braaatik** [comment]
No, its without --enable-cassert:
```
./configure --with-python --with-libxml --prefix=/usr/local/gpdb
```

**Comment by braaatik** [comment]
I looked into the source code and stack trace more carefully. Agree with you, its looks like the same

**Comment by JunfengYang** [comment]
Finally, find the root cause.
The main root cause is that if a LOV has multiple bitmap pages, and the index insert tries to insert a tid into a compressed word on a full bitmap page(Let's call the page `PAGE_FULL`).
Then it'll try to find free space on next bitmap page(Let's call the page `PAGE_NEXT`) and rearrange the words and copy extra words into the next bitmap page.

So when the above insertion happens, imagine below case:
1. Query on bitmap: A query starts and reads all bitmap pages to `PAGE_FULL`, increase next tid to fetch, release lock after reading each page.
2. Concurrent insert: insert a tid into `PAGE_FULL` cause expand compressed words to new words, and rearrange words into `PAGE_NEXT`.
3. Query on bitmap: fetch `PAGE_NEXT` and expect the first tid in it should equal the saved next tid. But actually `PAGE_NEXT` now contains words used to belong in `PAGE_FULL`.
This causes the real next tid less than the expected next tid. But our scan keeps increasing the wrong tid. And then this leads to a wrong result.

Will check whether the master branch has this issue or not and work on a fix.

**Comment by JunfengYang** [comment]
Normally the insert `tids` are inserted at the end of each LOV, and can not trigger the but.
There are some exceptions:
1. For heap table, after `vacuum` or during `vacuum full`.

2. For AO tables, this happens more frequently, since the tid are composed of seg file number and row number.
`vacuum full` will recreate index in another seg file number and current session's seg file number for an ao table
could be smaller than the recreated one. So insert new tuples in index may case update in bitmap page.
Also if the session changed, it'll choose another seg file number for the session which also may cause an update on bitmap page
when inserting new tuples.

**Comment by braaatik** [comment]
Can we hope this will be fixed for GP 6X?

**Comment by JunfengYang** [comment]
The fix is merged into master 22c1ec81c1ce723c3760e70da85cf4742cffcfc9 and 6X 3420e7701acdf176126c8ef7ece19070862336ce.
